refactor(Clients): Modularize BaseClient and improve code quality

mgajek-cern opened the pull request:
Closes #8057

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -92,148 +98,295 @@ class BaseClient:
         Parameters
         ----------
-        rucio_host :
+        rucio_host : Optional[str]
```
We don't type in the docstrings - https://rucio.cern.ch/documentation/developer/dev_style_guide

mgajek-cern replied:
Thanks for the feedback. Addressed in subsequent commits

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -92,148 +98,295 @@ class BaseClient:
+        
+        self.list_hosts = [self.host]
 
+    def _get_config_value(self, section: str, option: str) -> str:
```
Both of these config functions introduce unnecessary complexity, they don't handle every config failure so it just makes it harder to know if you should be using them or a standard config get. I would recommend just handling the exceptions for each parameter.

mgajek-cern replied:
I understand the concern, but these helpers eliminate duplicated exception handling across 10+ config reads in the constructor. They make the distinction between required and optional config explicit at the call site.

Both handle the exceptions relevant to their purpose. `_get_config_value()` wraps `NoOptionError/NoSectionError` in `MissingClientParameter` with better context for required values. `_get_optional_config()` returns `None` for missing optional values.

For `trace_host`, the helper makes intent clearer:

```python
self.trace_host = self._get_optional_config('trace', 'trace_host') or self.host
```

This is standard DRY. If you see specific failure cases these miss or have suggestions to improve them, I'm happy to discuss.

voetberg replied:
Right now it doesn't handle the config_get_{type} cases, or the rules to get defaults, or raise or not raise exceptions. I would argue it's better to just let the config_get raise a `ConfigNotFound` error if you care about readability

voetberg replied:
For your example, instead I would do
```
self.trace_host = config_get('trace', 'trace_host', default=self.host, raise_exception=False)
```

mgajek-cern replied:
Removed redundant helper methods in favor of `config_get(...)`

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -92,148 +98,295 @@ class BaseClient:
+        ClientProtocolNotSupported
+            If URL scheme is not in allowed list
+        """
+        self._validate_url_scheme(self.host, ['http', 'https'])
```
Wouldn't it be easier to just return the scheme from the validation here? You're just doing the urlparse operation twice. You can change the name to _get_valid_url_scheme for clarity

mgajek-cern replied:
Addressed in subsequent commits

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -617,128 +770,304 @@ def __refresh_token_oidc(self) -> bool:
-        # requesting authorization URL specific to the user & Rucio OIDC Client
-        self.logger.debug("Initial auth URL request headers %s to files" % str(headers))
+        
+        self.logger.debug(f"Initial auth URL request headers {headers}")
```
@rdimaio Do you have an opinion on using f strings in log messages? I remember a conversation we had on such a topic but don't remember the outcome

rdimaio replied:
More info here https://docs.astral.sh/ruff/rules/logging-f-string/, but basically it's best to avoid f-strings in log messages

mgajek-cern replied:
Addressed in subsequent commit.

Has been discussed toroughly here:  https://github.com/PyCQA/pylint/issues/2395. Summary: Performance tests show f-strings and %-formatting are virtually identical in practice, only about 1% difference even in million-iteration tests. Since Rucio uses Python ≥3.9, f-strings could be the modern, idiomatic choice for string formatting. Many teams prefer f-strings for readability and accept the minor trade-off.

voetberg commented on lib/rucio/client/baseclient.py:
I think this introduces a ton of code that is just for oidc and isn't marked as such. I would recommend moving the oidc auth things to their own class, and using it if the user selects oidc auth. `BaseClient` would become unmaintainable if this much functionality is introduced for multiple types of auth.

mgajek-cern replied:
I understand the concern, but this PR refactors existing OIDC code that was already in `BaseClient`. It doesn't add new functionality, just breaks down the 150+ line monolithic method into testable units per #8057.
Extracting auth types into separate strategy classes is a good architectural improvement, but it's different scope. This PR makes the existing code maintainable. We could tackle auth class extraction as a follow-up PR.

voetberg replied:
I have been wanting to do that for a long while so I'm using this as an opportunity to push my agenda haha, yes I think that's a good idea. See https://github.com/rucio/rucio/issues/7038 (Closed at the time because of other oidc client work that took priority)

mgajek-cern replied:
This PR just cleans up the existing OIDC code by breaking it into smaller, easier-to-maintain pieces. It doesn’t change how anything works, but it makes the code more manageable. That should make it easier to do a bigger cleanup like the one suggested in #7038 later on, where each auth type gets its own class.

Geogouz commented:
I'd like to review this. Can you rebase, resolving the conflict?

Geogouz commented on tests/test_baseclient_helpers.py:
```diff
@@ -0,0 +1,420 @@
+            elif option == 'trace_host':
+                if not raise_exception:
+                    return default
+                raise NoSectionError(section)
```
Shouldn't this be imported?

Geogouz replied:
`NoSectionError`

mgajek-cern replied:
Addressed in subsequent commit

Geogouz commented:
Test to fix pending

mgajek-cern commented:
> Test to fix pending

Addressed previous CI findings and rebased onto upstream/master

Geogouz commented on lib/rucio/client/baseclient.py:
```diff
@@ -131,110 +149,215 @@ def __init__(self,
-        rucio_scheme_allowed = ['http', 'https']
-        auth_scheme_allowed = ['http', 'https']
+        self.auth_token_file_path, self.token_exp_epoch_file, self.token_file, self.token_path = self._get_auth_tokens()
+        self.__authenticate()
```
The order here is `defaults` → `OIDC/account/TLS setup` → `token discovery` → `__authenticate()` → and finally `_configure_request_retries()`. And the authentication routines use `_send_request`, which guards network retries with `self.request_retries` but this will always use the hardcoded default (3) (even if the operator configured a higher value in `client.request_retries`). I think `_configure_request_retries()` should better run before any network I/O.

mgajek-cern replied:
This is a latent bug in the original code that the refactor exposed more clearly. 
Fixed in subsequent commit. Request retries now configured before authentication.

Geogouz commented on lib/rucio/client/baseclient.py:
```diff
@@ -243,147 +366,161 @@ def __init__(self,
+                creds['oidc_username'] = config_get('client', 'oidc_username', False, None)
+            if 'oidc_password' not in creds or creds['oidc_password'] is None:
+                creds['oidc_password'] = config_get('client', 'oidc_password', False, None)
+        if 'oidc_scope' not in creds or creds['oidc_scope'] == 'openid profile':
```
Should we guard against the cases where `oidc_scope` is None (which not sure how probable this is because CLI/config defaults keep it non-null for the moment)? For example later calls such as `_auto_authorize_oidc_scopes` would raise `AttributeError`.

mgajek-cern replied:
Current code already ensures oidc_scope is never None via default fallback ('openid profile'). 

---

We could make it more explicit if you think it's worth the defensive code. Eventually could be addressed in a subsequent issue. Thoughts?

voetberg replied:
A user could manually set `oidc_scope` as None and this check would skip it, I think it is worth defensively protecting against it. Good catch @Geogouz

mgajek-cern replied:
Addressed

Geogouz commented on lib/rucio/client/baseclient.py:
```diff
@@ -131,110 +149,215 @@ def __init__(self,
+        """
         try:
             self.request_retries = config_get_int('client', 'request_retries')
         except (NoOptionError, ConfigNotFound):
```
Should we maybe include here also `NoSectionError`?

mgajek-cern replied:
Considered in subsequent commit. Added NoSectionError to exception handling for consistency.

Geogouz commented on lib/rucio/client/baseclient.py:
```diff
@@ -243,147 +366,161 @@ def __init__(self,
     def _get_auth_tokens(self) -> tuple[Optional[str], str, str, str]:
-        # if token file path is defined in the rucio.cfg file, use that file. Currently this prevents authenticating as another user or VO.
+        """Get authentication token file paths."""
         auth_token_file_path = config_get('client', 'auth_token_file_path', False, None)
```
Maybe we could make use of `_expand_path()` here to make path resolving a bit more robust.

mgajek-cern replied:
Addressed in subsequent commit. Applied `_expand_path to` token_path for tilde expansion support.

Geogouz commented:
Placed couple comments here

**General note:** If you agree with those and proceed resolving them (with tests passing), please remove the `changes-required` label so it is easy for reviewers to notice. Also, if you need to discuss something, you could also place the label `Discussion needed`. And the same holds with any other PR (whether I am assigned to that as a reviewer or not) as I am tracking that label too.

mgajek-cern commented:
> Placed couple comments here
> 
> **General note:** If you agree with those and proceed resolving them (with tests passing), please remove the `changes-required` label so it is easy for reviewers to notice. Also, if you need to discuss something, you could also place the label `Discussion needed`. And the same holds with any other PR (whether I am assigned to that as a reviewer or not) as I am tracking that label too.

Thanks for the thorough review. All addressed:

1. Request retries: Fixed ordering bug
2. OIDC scope: Added explicit None guard
3. NoSectionError: Added for consistency  
4. _expand_path: Applied to token paths

Removing `changes-required` label.

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -60,16 +61,32 @@
 MAX_RETRY_BACK_OFF_SECONDS = 10
 
+# OIDC authentication constants
+OIDC_POLLING_TIMEOUT_SECONDS = 180  # 3 minutes
```
Should these constants instead live in `common/constants`?

mgajek-cern replied:
Currently, these constants aren’t reused elsewhere. Let’s keep them in `BaseClient` for now. If we find the need to share them in the future, we can move them to `common/constants.py` in a separate PR.

Geogouz commented:
I will give this later today another check. Can you rebase in case needed?

mgajek-cern commented:
> I will give this later today another check. Can you rebase in case needed?

Great. Rebased onto master

Geogouz reviewed (dismissed):
Can't see something clearly problematic any more. The refactor does still leave `BaseClient` as a large, multi‑concern class, but I find the extracted helper methods to be a step in the right direction. Maybe others want to give it also a reviewing, but at least from my side, it gets a green light.

dchristidis reviewed (changes requested):
This PR achieves too many things in a single commit. Kindly split it into multiple, purposeful commits.

mgajek-cern commented:
> This PR achieves too many things in a single commit. Kindly split it into multiple, purposeful commits.

@dchristidis. I kindly disagree with splitting this into multiple commits.

This is a single, cohesive refactoring addressing #8057: modularizing BaseClient's constructor and OIDC flow. The changes are tightly coupled: constructor helpers, OIDC flow extraction and code quality improvements all work together to transform a monolithic class into a modular one.

Splitting would create incomplete intermediate states and obscure the unified "before → after" narrative. The commit message clearly documents what was extracted and improved.

After ~4 months of review with approvals from @Geogouz and @voetberg, this follows conventional commit standards for single-purpose refactors. Happy to address specific technical concerns, but the commit structure is appropriate.

dchristidis commented:
> Splitting would create incomplete intermediate states and obscure the unified "before → after" narrative. The commit message clearly documents what was extracted and improved.

Please take a moment to explain how the following, for example, is an integral part of the one commit and both cannot and should not be split.

> Standardize string formatting to f-strings and replace print() calls with proper logging.

mgajek-cern commented:
> > Splitting would create incomplete intermediate states and obscure the unified "before → after" narrative. The commit message clearly documents what was extracted and improved.
> 
> Please take a moment to explain how the following, for example, is an integral part of the one commit and both cannot and should not be split.
> 
> > Standardize string formatting to f-strings and replace print() calls with proper logging.

Appreciate the revieq. However, I kindly disagree that f-string conversion and logging improvements should be split from this refactoring.

Why they're integral:

When extracting 15+ helper methods from a 500-line monolithic __init__, standard refactoring practice is to clean up each extracted method's internals. Copying legacy % formatting and print() statements into new helpers would introduce technical debt immediately after creation. Splitting would require alternating extraction and cleanup commits for each helper:

```
Extract _setup_session() (with old % formatting)
Modernize _setup_session() (convert to f-strings)
Extract _configure_hosts() (with old % formatting)
Modernize _configure_hosts() (convert to f-strings)
...repeated 15+ times across the same file.
```

This creates:

- Massive code churn (30+ commits touching identical lines)
- Incomplete intermediate states (half the class modern, half legacy)
- Inconsistent codebase between commits

Proposed solution:
- I can make the commit message more explicit about all changes

```
refactor(Clients): Modularize BaseClient and standardize code quality

Break down the monolithic __init__ and __get_token_oidc methods into
focused helper methods, improving readability, testability and
maintainability. Code quality improvements are applied during extraction
per standard refactoring practice.

WHAT CHANGED:

Constructor modularization:
- Extract _setup_session, _configure_hosts, _configure_account_and_vo,
  _validate_and_configure_tls, _configure_request_retries

OIDC flow modularization:
- Extract _handle_oidc_auto_flow, _handle_oidc_polling_flow,
  _handle_oidc_manual_code_flow, _request_oidc_auth_url,
  _finalize_oidc_token, _auto_authorize_oidc_scopes

Code quality and style improvements in extracted methods:
- Standardize string formatting to f-strings (% → f"")
- Replace print() calls with proper logging (self.logger)
- Add missing type hints (parameters and return types)
- Replace bare exception handlers with explicit ones (Exception → OSError)
- Extract magic header strings into constants (HEADER_RUCIO_AUTH_TOKEN, etc.)
- Extract OIDC constants (OIDC_POLLING_TIMEOUT_SECONDS, OIDC_MAX_CODE_ATTEMPTS)
- Remove redundant file existence checks
- Improve error messages and docstrings for clarity
- Add comprehensive unit tests for extracted helpers

Type system improvements:
- Update _get_exception to accept Mapping[str, str] for better compatibility

WHY TOGETHER:

When extracting 15+ helper methods from a 500-line __init__, each new method
is cleaned up during extraction. Splitting extraction from style improvements would
require 30+ alternating commits (extract → modernize → extract → modernize)
touching the same lines repeatedly, creating:
- Massive code churn (same file modified 30+ times)
- Incomplete intermediate states (mixing modern and legacy styles)
- Inconsistent codebase between commits

Modern Python style (f-strings, type hints, explicit logging) is applied to
extracted code to avoid introducing technical debt in newly created methods.

IMPACT:

No functional changes. Pure refactor with better modularity, readability,
easier debugging and maintenance and improved testability through smaller,
isolated helper methods.

Closes: #8057
```

This maintains atomic revertability while documenting all changes explicitly. If needed, the the commit message can be revised.

dchristidis commented:
> However, I kindly disagree

Then I fear that we are at an impasse.

> standard refactoring practice is to clean up each extracted method's internals.

Refactoring is done iteratively. The more changes applied in one go, the more difficult to keep track, and the higher the risk of introducing a regression.

> Copying legacy % formatting and print() statements into new helpers would introduce technical debt immediately after creation.

I do not follow your argument, given that you can always make the clean-ups before you start moving things around.

> Splitting would require alternating extraction and cleanup commits for each helper:

How can you arrive at such a conclusion? I never asked for this.

> This maintains atomic revertability while documenting all changes explicitly.

Historically there has been little need to revert a commit. Optimising for it comes at the cost of reviewing (actively done right now) and future debugging (done very often). Documenting in the commit is not enough if it takes effort to associate the change in the code with the description.

mgajek-cern commented:
> > However, I kindly disagree
> 
> Then I fear that we are at an impasse.
> 
> > standard refactoring practice is to clean up each extracted method's internals.
> 
> Refactoring is done iteratively. The more changes applied in one go, the more difficult to keep track, and the higher the risk of introducing a regression.
> 
> > Copying legacy % formatting and print() statements into new helpers would introduce technical debt immediately after creation.
> 
> I do not follow your argument, given that you can always make the clean-ups before you start moving things around.
> 
> > Splitting would require alternating extraction and cleanup commits for each helper:
> 
> How can you arrive at such a conclusion? I never asked for this.
> 
> > This maintains atomic revertability while documenting all changes explicitly.
> 
> Historically there has been little need to revert a commit. Optimising for it comes at the cost of reviewing (actively done right now) and future debugging (done very often). Documenting in the commit is not enough if it takes effort to associate the change in the code with the description.

This PR implements the approved solution from issue #8057, which explicitly requested both structural refactoring and code quality improvements together (see 'Non-Idiomatic Patterns' section). Tough I need to admit that the title of the issue might need to be adjusted possibly for clarity.

Lets discuss your concerns in our meeting next week.

dchristidis commented:
I made a rudimentary attempt to split you commit. Maybe this will help see my point of view.

https://github.com/dchristidis/rucio/tree/patch-8057-refactor-base-client

The top of the two branches is the same. Same for the base.

From one commit:
```
 lib/rucio/client/baseclient.py   | 1288 +++++++++++++++++++-----------
```

To multiple commits:
```
 lib/rucio/client/baseclient.py | 53 ++++++++++++++++++------------------------
 lib/rucio/client/baseclient.py | 52 ++++++++++++++++++++----------------------
 lib/rucio/client/baseclient.py | 86 ++++++++++++++++++++----------------------
 lib/rucio/client/baseclient.py   | 42 ++++++++++++++++++++--------------------
 lib/rucio/client/baseclient.py | 138 +++++++++++++++++++++++++----------------
 lib/rucio/client/baseclient.py | 194 +++++++++++++++++++++++++++++------------
 lib/rucio/client/baseclient.py | 75 ++++++++++++++++++++++++++----------------
 lib/rucio/client/baseclient.py | 50 ++++++++++++++++++++----------------------
 lib/rucio/client/baseclient.py | 59 ++++++++++++++++++++++++++----------------
 lib/rucio/client/baseclient.py | 49 ++++++++++++++++++++++++++----------------
 lib/rucio/client/baseclient.py | 185 ++++++++++++++++++++++++++---------------
 lib/rucio/client/baseclient.py | 38 +++++++++++++++++++++++++++-----------
 lib/rucio/client/baseclient.py | 30 ++++++++++++++++++++++--------
 lib/rucio/client/baseclient.py | 30 ++++++++++++++----------------
 lib/rucio/client/baseclient.py | 16 +++++++---------
 lib/rucio/client/baseclient.py | 29 ++++++++++++-----------------
 lib/rucio/client/baseclient.py | 172 +++++++++++++++++++++--------------------
 lib/rucio/client/baseclient.py | 68 +++++++++++++++---------------------------
 lib/rucio/client/baseclient.py | 46 ++++++++++++++++++++++++++++--------------
 lib/rucio/client/baseclient.py | 46 ++++++++++++++++++++++++++++--------------
 lib/rucio/client/baseclient.py | 184 ++++++++++++++++++++++++++++++-----------
```

In the second case, they are much more manageable to review: each one (mostly) does one thing, and the size allows keeping all of it in one’s mind.

mgajek-cern commented:
> I made a rudimentary attempt to split you commit. Maybe this will help see my point of view.
> 
> https://github.com/dchristidis/rucio/tree/patch-8057-refactor-base-client
> 
> The top of the two branches is the same. Same for the base.
> 
> From one commit:
> 
> ```
>  lib/rucio/client/baseclient.py   | 1288 +++++++++++++++++++-----------
> ```
> 
> To multiple commits:
> 
> ```
>  lib/rucio/client/baseclient.py | 53 ++++++++++++++++++------------------------
>  lib/rucio/client/baseclient.py | 52 ++++++++++++++++++++----------------------
>  lib/rucio/client/baseclient.py | 86 ++++++++++++++++++++----------------------
>  lib/rucio/client/baseclient.py   | 42 ++++++++++++++++++++--------------------
>  lib/rucio/client/baseclient.py | 138 +++++++++++++++++++++++++----------------
>  lib/rucio/client/baseclient.py | 194 +++++++++++++++++++++++++++++------------
>  lib/rucio/client/baseclient.py | 75 ++++++++++++++++++++++++++----------------
>  lib/rucio/client/baseclient.py | 50 ++++++++++++++++++++----------------------
>  lib/rucio/client/baseclient.py | 59 ++++++++++++++++++++++++++----------------
>  lib/rucio/client/baseclient.py | 49 ++++++++++++++++++++++++++----------------
>  lib/rucio/client/baseclient.py | 185 ++++++++++++++++++++++++++---------------
>  lib/rucio/client/baseclient.py | 38 +++++++++++++++++++++++++++-----------
>  lib/rucio/client/baseclient.py | 30 ++++++++++++++++++++++--------
>  lib/rucio/client/baseclient.py | 30 ++++++++++++++----------------
>  lib/rucio/client/baseclient.py | 16 +++++++---------
>  lib/rucio/client/baseclient.py | 29 ++++++++++++-----------------
>  lib/rucio/client/baseclient.py | 172 +++++++++++++++++++++--------------------
>  lib/rucio/client/baseclient.py | 68 +++++++++++++++---------------------------
>  lib/rucio/client/baseclient.py | 46 ++++++++++++++++++++++++++++--------------
>  lib/rucio/client/baseclient.py | 46 ++++++++++++++++++++++++++++--------------
>  lib/rucio/client/baseclient.py | 184 ++++++++++++++++++++++++++++++-----------
> ```
> 
> In the second case, they are much more manageable to review: each one (mostly) does one thing, and the size allows keeping all of it in one’s mind.

Based on the [Rucio Contributing Guidelines](https://rucio.github.io/documentation/contributing/)(https://rucio.github.io/documentation/contributing/), this should be **1-2 commits**:

**Recommended, e.g.:**
```
refactor(Clients): Modularize BaseClient and improve code quality

Extract helper methods from monolithic __init__ and __get_token_oidc.
Apply code quality improvements (f-strings, type hints, logging) during
extraction per Python best practices.

Closes: #8057
```

**Why:** The guidelines say "multiple commits should be squashed" unless "independent, well-defined." Their own example shows 60+ file modernization in **one commit**. The 22-commit approach contradicts documented standards.

Optional 2nd commit for tests if needed. Let's discuss this professionally in our meeting next week

dchristidis commented:
The reasoning behind squashing commits is to avoid merging ‘bad states’ into the history. A typical example would be someone making a commit, realising that the tests fail as a result of their commit, and then adding another commit to fix their original commit.

You can have a commit that spans many files, as long as you’re doing exactly one thing. For example, replacing uses %-formatting with f-strings across the entire codebase would be acceptable.

Incidentally, over the weekend I attended [a talk](https://fosdem.org/2026/schedule/event/3VNNBK-efficient-git-for-high-stakes/) which advocates for what I’m describing. I’ll present a summary of it at the Thursday Rucio meeting and later update the contributing guide accordingly.

mgajek-cern committed:
refactor(Clients): Extract magic strings and OIDC constants

Extract HTTP header names into constants (HEADER_RUCIO_AUTH_TOKEN, etc.)
and OIDC configuration values (OIDC_POLLING_TIMEOUT_SECONDS, etc.).

This improves maintainability and reduces magic strings throughout the code.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Add type hints to BaseClient methods

Add comprehensive type hints to method signatures and variables.
Update _get_exception to accept Mapping[str, str] for better type compatibility.
Import Mapping from collections.abc for runtime usage.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Replace string formatting with f-strings

Modernize string formatting throughout BaseClient by replacing
%-formatting and .format() calls with f-strings.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Break down BaseClient.__init__ into helper methods

Extract initialization logic into focused helper methods:
- _setup_session: Initialize HTTP session and user agent
- _configure_hosts: Configure Rucio and auth server hosts
- _configure_account_and_vo: Configure account and VO with fallbacks
- _setup_oidc_config: Setup OIDC-specific configuration
- _validate_and_configure_tls: Validate URL schemes and configure TLS
- _configure_request_retries: Configure request retry count

This improves readability and testability without changing functionality.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Break down _get_creds into credential-specific methods

Extract credential population logic into separate methods:
- _populate_oidc_creds
- _populate_userpass_creds
- _populate_x509_creds
- _populate_x509_proxy_creds
- _populate_ssh_creds

Improves separation of concerns and makes testing easier.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Break down __get_token_oidc into flow-specific handlers

Split OIDC authentication into distinct flow handlers:
- _request_oidc_auth_url: Request authorization URL from server
- _build_oidc_request_headers: Build HTTP headers for OIDC request
- _handle_oidc_polling_flow: Handle polling-based authentication
- _handle_oidc_manual_code_flow: Handle manual code entry
- _handle_oidc_auto_flow: Handle automatic authentication
- _auto_authorize_oidc_scopes: Auto-authorize scope requests
- _finalize_oidc_token: Extract and store OIDC token

Makes complex OIDC logic more maintainable and testable.

Issue: #8057

mgajek-cern committed:
refactor(Clients): Improve error handling and logging

Replace bare exception handlers with explicit exception types.
Replace print() calls with proper logger calls.
Improve error messages for clarity.
Remove redundant file existence checks.

Issue: #8057

mgajek-cern committed:
docs(Clients): Improve docstrings and add parameter documentation

Add comprehensive docstrings to all new helper methods following
NumPy/Google style guide conventions. Document parameters, return
values, and exceptions.

Issue: #8057

mgajek-cern committed:
test(Clients): Add unit tests for BaseClient helper methods

Add comprehensive unit tests for extracted helper methods:
- Session setup and configuration
- Host and account/VO configuration
- URL scheme validation
- CA certificate discovery
- All OIDC authentication flows

Improves test coverage and validates refactoring.

Issue: #8057

mgajek-cern committed:
fix(Clients): Add type ignore comment for replica state access

Add type ignore comment to suppress mypy error when accessing replica
state dictionary. The replica state returned by list_replicas has a
dynamic structure that mypy cannot fully type-check.

Issue: #8057

mgajek-cern committed:
style(Clients): Clean up formatting and fix type conversion

Remove unnecessary comments, fix inconsistent formatting, and improve
code readability throughout BaseClient. Also fix type error where token
expiration epoch was assigned as string instead of int.

Changes:
- Fix typo: "Lost of hosts" → "List of hosts"
- Remove extra blank line after class declaration
- Add blank lines for PEP 8 compliance in _get_exception
- Replace 'else: return' with direct return (more Pythonic)
- Remove obsolete Python 2/3 compatibility comments
- Remove unnecessary pylint disable comments
- Fix multi-line dict formatting in __get_token_userpass
- Remove duplicate unreachable 'return False' in __refresh_token_oidc
- Fix type error: convert token_exp_epoch to int in __refresh_token_oidc
- Clean up multi-line string formatting in __refresh_token_oidc
- Fix __get_token_saml return pattern for consistency
- Standardize debug message capitalization
- Remove implementation detail comments

Issue: #8057

mgajek-cern commented:
> The reasoning behind squashing commits is to avoid merging ‘bad states’ into the history. A typical example would be someone making a commit, realising that the tests fail as a result of their commit, and then adding another commit to fix their original commit.
> 
> You can have a commit that spans many files, as long as you’re doing exactly one thing. For example, replacing uses %-formatting with f-strings across the entire codebase would be acceptable.
> 
> Incidentally, over the weekend I attended [a talk](https://fosdem.org/2026/schedule/event/3VNNBK-efficient-git-for-high-stakes/) which advocates for what I’m describing. I’ll present a summary of it at the Thursday Rucio meeting and later update the contributing guide accordingly.

I've restructured the PR into 11 focused commits following your feedback and todays meeting. Each commit:

- Does exactly one thing (following your f-string example)
- Is independently reviewable (few hundred lines)
- Passes all tests
- Can be reverted without breaking others

The progression is:
- 1. Code modernization (constants, types, f-strings, logging)
- 2. Structural refactoring (constructor, credentials, OIDC flows)
- 3. Quality improvements (errors, docs)
- 4. Comprehensive tests
- 5. Style concerns

This maintains the cohesive refactoring from #8057 while addressing your reviewability concerns. Each commit is bisectable and tells a clear story of the transformation.

voetberg commented on lib/rucio/client/baseclient.py:
Should this string be similarly extracted in `c235bfc789e8adf31cd957b0874e6c7ffa3f7dea`

voetberg commented on lib/rucio/client/baseclient.py:
Needs annotations

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -108,17 +109,19 @@ class BaseClient:
-                 user_agent: Optional[str] = 'rucio-clients',
-                 vo: Optional[str] = None,
-                 logger: 'Logger' = LOG) -> None:
+    def __init__(
```
This is a style change in 770148, please move to style c052f032

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -391,7 +395,12 @@ def _get_creds(self, creds: Optional[dict[str, Any]]) -> dict[str, Any]:
         return creds
 
-    def _get_exception(self, headers: dict[str, str], status_code: Optional[int] = None, data=None) -> tuple[type[exception.RucioException], str]:
+    def _get_exception(
```
Another style change in 7701428

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -162,15 +162,15 @@ def __init__(
                 self.host = config_get('client', 'rucio_host')
         except (NoOptionError, NoSectionError) as error:
-            raise MissingClientParameter('Section client and Option \'%s\' cannot be found in config file' % error.args[0])
+            raise MissingClientParameter(f"Section client and Option '{error.args[0]}' cannot be found in config file")
```
Putting the fstring in the message renders the error strangely, do 
```
msg = ....
raise Error(msg)
```
instead please

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -162,15 +162,15 @@ def __init__(
                 self.auth_host = config_get('client', 'auth_host')
         except (NoOptionError, NoSectionError) as error:
-            raise MissingClientParameter('Section client and Option \'%s\' cannot be found in config file' % error.args[0])
+            raise MissingClientParameter(f"Option '{error.args[0]}' cannot be found in config file")
```
Re: https://github.com/rucio/rucio/pull/8081/changes/1fbdba6a4641e5d6b32047d2c44cdeb748d9b213#r2878980988

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -274,25 +275,24 @@ def _get_auth_tokens(self) -> tuple[Optional[str], str, str, str]:
             if 'RUCIO_AUTH_TYPE' in environ:
                 if environ['RUCIO_AUTH_TYPE'] not in ['userpass', 'x509', 'x509_proxy', 'gss', 'ssh', 'saml', 'oidc']:
-                    raise MissingClientParameter('Possible RUCIO_AUTH_TYPE values: userpass, x509, x509_proxy, gss, ssh, saml, oidc, vs. ' + environ['RUCIO_AUTH_TYPE'])
+                    raise MissingClientParameter(f"Possible RUCIO_AUTH_TYPE values: userpass, x509, x509_proxy, gss, ssh, saml, oidc, vs. {environ['RUCIO_AUTH_TYPE']}")
```
Re: https://github.com/rucio/rucio/pull/8081/changes/1fbdba6a4641e5d6b32047d2c44cdeb748d9b213#r2878980988

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -150,130 +150,38 @@ def __init__(
-        except (NoOptionError, NoSectionError, ConfigNotFound):
-            self.trace_host = self.host
-            self.logger.debug('No trace_host passed. Using rucio_host instead')
+        self._setup_session(user_agent)
```
This is totally up for debate by my personal preference here would be to not sequester all the class in these helper functions, even if they are set in the helper function (it just makes it harder to track down what attrs exist in the future). 

I would prefer if you set the attrs used here as `None` before you set them in the helper functions.

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+        """
+        self.session = Session()
+        self.user_agent = f"{user_agent}/{version.version_string()}"
+        if sys.argv:
```
Strictly speaking this is a logic change and should be a different commit - The original code is: 

```
        sys.argv[0] = sys.argv[0].split('/')[-1]
        self.script_id = '::'.join(sys.argv[0:2])
```

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+
+        self.trace_host = config_get('trace', 'trace_host', raise_exception=False, default=self.host)
+        if self.trace_host == self.host:
+            self.logger.debug('No trace_host configured. Using rucio_host instead')
```
Also another logic change. Original code is:

```
        try:
            self.trace_host = config_get('trace', 'trace_host')
        except (NoOptionError, NoSectionError, ConfigNotFound):
            self.trace_host = self.host
            self.logger.debug('No trace_host passed. Using rucio_host instead')

```
The debug would be missed if the user manually sets a trace-host that is the same as the original host, but the config fails to load it.

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+        vo :
+            Virtual Organization name
+        """
+        self.account = (
```
Missing debug statement  -             `self.logger.debug('No account passed. Trying to get it from the RUCIO_ACCOUNT environment variable or the config file.')`

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+                or config_get('client', 'vo', raise_exception=False, default=None)
+                or DEFAULT_VO
+            )
+            if self.vo == DEFAULT_VO:
```
Not strictly true logic. The user can pass the default VO and get this debug statement.

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+            self.logger.debug('X509_CERT_DIR not defined. Trying config file.')
+            try:
+                ca_cert = _expand_path(config_get('client', 'ca_cert'))
+            except (NoOptionError, NoSectionError, ConfigNotFound):
```
Removes the different debug statement -

```
                except (NoOptionError, NoSectionError):
                    self.logger.debug('No ca_cert found in configuration. Falling back to Mozilla default CA bundle (certifi).')
                    self.ca_cert = True
                except ConfigNotFound:
                    self.logger.debug('No configuration found. Falling back to Mozilla default CA bundle (certifi).')
                    self.ca_cert = True
```

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1069,3 +977,199 @@ def __authenticate(self) -> None:
+        """
+        try:
+            self.request_retries = config_get_int('client', 'request_retries')
+        except (NoOptionError, NoSectionError, ConfigNotFound):
```
Logic change by adding the `NoSectionError` - add this to a different commit

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -204,102 +204,22 @@ def _get_auth_type(self, auth_type: Optional[str]) -> str:
         return auth_type
 
     def _get_creds(self, creds: Optional[dict[str, Any]]) -> dict[str, Any]:
+        """Get credentials from parameter or config file."""
```
Documentation change, different commit please

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1173,3 +1093,82 @@ def _configure_request_retries(self) -> None:
+    def _populate_x509_creds(self, creds: dict[str, Any]) -> None:
+        """Populate X509 credentials from config."""
+        if 'client_cert' not in creds or creds['client_cert'] is None:
+            creds['client_cert'] = environ.get("RUCIO_CLIENT_CERT") or config_get('client', 'client_cert')
```
`RUCIO CLIENT CERT` should probably be considered a magic string

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1173,3 +1093,82 @@ def _configure_request_retries(self) -> None:
+            )
+
+    def _populate_x509_proxy_creds(self, creds: dict[str, Any]) -> None:
+        """Populate X509 proxy credentials from config."""
```
Removes the comments that may be useful for future documentation

```
                # rucio specific configuration takes precedence over GSI logic
                # environment variables take precedence over config values
                # So we check in order:
                # RUCIO_CLIENT_PROXY env variable
                # client.client_x509_proxy rucio cfg variable
                # X509_USER_PROXY env variable
                # /tmp/x509up_u`id -u` if exists
```

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1172,3 +1081,209 @@ def _populate_ssh_creds(self, creds: dict[str, Any]) -> None:
+        self.logger.debug("Response headers %s and text %s", oidc_auth_res.headers, oidc_auth_res.text)
+
+        if HEADER_RUCIO_OIDC_AUTH_URL not in oidc_auth_res.headers:
+            self.logger.error("Failed to get AuthN/Z URL from Rucio Auth Server. Check scope, audience, or issuer configuration.")
```
It may be cleaner to just raise the `cannot authenticate` error here instead of in the main function

voetberg replied:
This goes for all the `return False`s and `return None`s that just go over to the cannot auth errors

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -1172,3 +1081,209 @@ def _populate_ssh_creds(self, creds: dict[str, Any]) -> None:
         if not os.path.isfile(creds["ssh_private_key"]):
             raise CannotAuthenticate(f'Provided ssh private key {creds["ssh_private_key"]!r} does not exist')
+
+    def _request_oidc_auth_url(self) -> Optional[str]:
```
Sort of a nick-pick, but it may be better to have these functions under the oidc auth flow to avoid a lot of extra up/down scrolling

voetberg commented on lib/rucio/client/baseclient.py:
```diff
@@ -577,12 +576,10 @@ def __get_token_ssh(self) -> bool:
         headers = {}
 
         private_key_path = self.creds['ssh_private_key']
+
```
Nickpick - please remove this random new line it clutters the blame

mgajek-cern commented:
Given this PR has been in review for ~6  months with recurring cycles of new comments, would it make sense to either define a clear path to merge or close it?

We can do either. Just trying to avoid it stalling indefinitely.

bari12 commented:
Let's close this for now and re-visit it once we have more time. I think we need to handle this very much in the way with other L/XL issues where we discuss the changes with the component lead beforehand and make a step by step change.

mgajek-cern commented:
Thanks for the clarification. Let's proceed this way.

One thought for future larger contributions: would it be useful to add a short section in the [Contributing guidelines](https://rucio.github.io/documentation/contributing/) describing how to approach `L` / `XL` / `XXL` changes? 

For example:

- splitting larger changes into smaller, logically scoped PRs
- sequencing dependent refactors step-by-step
- using [stacked PRs](https://github.github.com/gh-stack/) (or similar workflows) where appropriate
- agreeing on an incremental merge plan early with the component lead

These approaches can help improve reviewability, reduce risk and avoid long-lived PRs. Even lightweight guidance could be helpful for future contributors
